### DISCUSSION TITLE
goto in action - The exception from the rule

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#10] [General]
# goto in action - The exception from the rule

A lot of developers will say never use goto. This keyword is bad as it promotes spaghetti code. But as so often in live the exception confirms the rules. Let's see where goto can increase the readability and maintainability.

https://steven-giesel.com/blogPost/a1de0393-e0f9-491e-8c26-8ec34da60d16

**Comment by wgebczyk** [comment]
What about:
```
private void SaveFile(string relativeOrAbsolutePath, PathKind pathKind)
{
    if (pathKind == PathKind.Relative)
    {
        relativeOrAbsolutePath = Path.Combine(currentPath, relativeOrAbsolutePath);
    }
    SaveFile(relativeOrAbsolutePath);
}
```

**Reply by linkdotnet** [reply]
> What about:
> 
> ```
> private void SaveFile(string relativeOrAbsolutePath, PathKind pathKind)
> {
>     if (pathKind == PathKind.Relative)
>     {
>         relativeOrAbsolutePath = Path.Combine(currentPath, relativeOrAbsolutePath);
>     }
>     SaveFile(relativeOrAbsolutePath);
> }
> ```

That would have the some outcome, exactly. Would fall under my first mentioned point: It is often more desirable to refactor the code. Still I wanted to show-case where `goto` might be a good choice (where it is not as simple as my selected example).

**Reply by wgebczyk** [reply]
A bit better example -- https://github.com/dotnet/runtime/blob/de078d1d078d1da6db4717d0e2618d12183f5c66/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L353

**Reply by linkdotnet** [reply]
> A bit better example -- https://github.com/dotnet/runtime/blob/de078d1d078d1da6db4717d0e2618d12183f5c66/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L353

Thanks nice example. I guess the sole reason they used `goto` here is performance (as a lot in this method).
Jumping around is rather cheap.
